protocols/perf/src/client/behaviour.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! [`NetworkBehaviour`] of the libp2p perf protocol.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    task::{Context, Poll},
+};
+
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::{
+    derive_prelude::ConnectionEstablished, dial_opts::DialOpts, ConnectionId, FromSwarm,
+    NetworkBehaviour, NetworkBehaviourAction, NotifyHandler, PollParameters, THandlerInEvent,
+    THandlerOutEvent,
+};
+
+use crate::{client::handler::Handler, RunParams, RunStats};
+
+#[derive(Debug)]
+pub enum Event {
+    Finished { stats: RunStats },
+}
+
+#[derive(Default)]
+pub struct Behaviour {
+    pending_run: HashMap<ConnectionId, RunParams>,
+    /// Queue of actions to return when polled.
+    queued_events: VecDeque<NetworkBehaviourAction<Event, THandlerInEvent<Self>>>,
+}
+
+impl Behaviour {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn perf(&mut self, server: Multiaddr, params: RunParams) {
+        let opts: DialOpts = server.into();
+        let connection_id = opts.connection_id();
+
+        self.pending_run.insert(connection_id, params);
+
+        // TODO: What if we are already connected?
+        self.queued_events
+            .push_back(NetworkBehaviourAction::Dial { opts });
```

Comment:
I've developed the opinion that protocols should not manage connections unless their sole purpose is to do so (like dcutr or kad).

Instead, I'd state it as a prerequisite that we are connected and use `NotifyHandler::Any`.

You can then establish the connection within the binary through `Swarm::dial`.

Reply:
:+1: I share the same opinion. Changed with 39694af0.